rust/lancedb/src/embeddings/bedrock.rs

```diff
@@ -0,0 +1,210 @@
+        }
+    }
+}
+
+impl FromStr for BedrockEmbeddingModel {
+    type Err = Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "titan-embed-text-v1" => Ok(Self::TitanEmbedding),
+            "cohere-embed-english-v3" => Ok(Self::CohereLarge),
+            _ => Err(Error::InvalidInput {
+                message: "Invalid model. Available models are: 'titan-embed-text-v1', 'cohere-embed-english-v3'".to_string()
+            }),
+        }
+    }
+}
+
+pub struct BedrockEmbeddingFunction {
+    model: BedrockEmbeddingModel,
+    client: BedrockClient,
+}
+
+impl BedrockEmbeddingFunction {
+    pub fn new(client: BedrockClient) -> Self {
+        Self {
+            model: BedrockEmbeddingModel::TitanEmbedding,
+            client,
+        }
+    }
+
+    pub fn with_model(client: BedrockClient, model: BedrockEmbeddingModel) -> Self {
+        Self { model, client }
+    }
+}
+
+impl EmbeddingFunction for BedrockEmbeddingFunction {
+    fn name(&self) -> &str {
+        "bedrock"
+    }
+
+    fn source_type(&self) -> Result<Cow<DataType>> {
+        Ok(Cow::Owned(DataType::Utf8))
+    }
+
+    fn dest_type(&self) -> Result<Cow<DataType>> {
+        let n_dims = self.model.ndims();
+        Ok(Cow::Owned(DataType::new_fixed_size_list(
+            DataType::Float32,
+            n_dims as i32,
+            false,
+        )))
+    }
+
+    fn compute_source_embeddings(&self, source: ArrayRef) -> Result<ArrayRef> {
+        let len = source.len();
+        let n_dims = self.model.ndims();
+        let inner = self.compute_inner(source)?;
+
+        let fsl = DataType::new_fixed_size_list(DataType::Float32, n_dims as i32, false);
+
+        let array_data = ArrayData::builder(fsl)
+            .len(len)
+            .add_child_data(inner.into_data())
+            .build()?;
+
+        Ok(Arc::new(FixedSizeListArray::from(array_data)))
+    }
+
+    fn compute_query_embeddings(&self, input: Arc<dyn Array>) -> Result<Arc<dyn Array>> {
+        let arr = self.compute_inner(input)?;
+        Ok(Arc::new(arr))
+    }
+}
+
+impl std::fmt::Debug for BedrockEmbeddingFunction {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BedrockEmbeddingFunction")
+            .field("model", &self.model)
+            // Skip client field as it doesn't implement Debug
+            .finish()
+    }
+}
+
+impl BedrockEmbeddingFunction {
+    fn compute_inner(&self, source: Arc<dyn Array>) -> Result<Float32Array> {
+        if source.is_nullable() {
+            return Err(Error::InvalidInput {
+                message: "Expected non-nullable data type".to_string(),
+            });
+        }
+
+        if !matches!(source.data_type(), DataType::Utf8 | DataType::LargeUtf8) {
+            return Err(Error::InvalidInput {
+                message: "Expected Utf8 data type".to_string(),
+            });
+        }
+
+        let mut builder = Float32Builder::new();
```

Comment:
You can pre-size the builder to avoid reallocations:
```suggestion
        let mut builder = Float32Builder::with_capacity(self.ndims() * source.len());
```